kstatus/wait/wait.go

```diff
@@ -0,0 +1,222 @@
+	Completed EventType = "Completed"
+
+	// The wait was stopped before all resources could reach the
+	// Current status.
+	Aborted EventType = "Aborted"
+)
+
+// EventResource contains information about the resource for which
+// a specific Event pertains.
+type EventResource struct {
+	// Identifier contains information that identifies which resource
+	// this information is about.
+	Identifier ResourceIdentifier
+
+	// Status is the latest status for the given resource.
+	Status status.Status
+
+	// Message is more details about the status.
+	Message string
+
+	// Error is set if there was a problem identifying the status
+	// of the resource. For example, if polling the cluster for information
+	// about the resource failed.
+	Error error
+}
+
+// WaitForStatus polls all the provided resources until all of them has
+// reached the Current status. Updates as resources change their status and
+// when the wait is either completed or aborted is returned through a channel.
+func (r *Resolver) WaitForStatus(ctx context.Context, resources []ResourceIdentifier, pollInterval time.Duration) <-chan Event {
+	eventChan := make(chan Event)
+
+	go func() {
+		ticker := time.NewTicker(pollInterval)
+		defer func() {
+			ticker.Stop()
+			close(eventChan)
+		}()
+
+		if len(resources) == 0 {
+			eventChan <- Event{
+				Type:            Completed,
+				AggregateStatus: status.CurrentStatus,
+				EventResource:   nil,
+			}
+			return
+		}
+
+		waitState := newWaitState(resources, r.statusComputeFunc)
+
+		for {
+			select {
+			case <-ctx.Done():
+				eventChan <- Event{
+					Type:            Aborted,
+					AggregateStatus: waitState.AggregateStatus(),
+				}
+				return
+			case <-ticker.C:
+				for id := range waitState.ResourceWaitStates {
+					identifier := id
+					u, err := r.fetchResource(ctx, &identifier)
+					eventResource, isNew := waitState.ResourceObserved(&identifier, u, err)
+					aggStatus := waitState.AggregateStatus()
+					if isNew {
+						eventChan <- Event{
+							Type:            ResourceUpdate,
+							AggregateStatus: aggStatus,
+							EventResource:   &eventResource,
+						}
+					}
+					if aggStatus == status.CurrentStatus {
+						eventChan <- Event{
+							Type:            Completed,
+							AggregateStatus: status.CurrentStatus,
+						}
+						return
+					}
+				}
+			}
+		}
+	}()
+
+	return eventChan
+}
+
+// fetchResource gets the resource given by the identifier from the cluster
+// through the client available in the Resolver. It returns the resource
+// as an Unstructured.
+func (r *Resolver) fetchResource(ctx context.Context, identifier ResourceIdentifier) (*unstructured.Unstructured, error) {
+	key := types.NamespacedName{Name: identifier.GetName(), Namespace: identifier.GetNamespace()}
+	u := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": identifier.GetAPIVersion(),
+			"kind":       identifier.GetKind(),
+		},
+	}
+	err := r.client.Get(ctx, key, u)
+	return u, err
```

Comment:
We should wrap this in errors so it has a stack trace that can be printed in debug mode.